### PR TITLE
DOC: soften wording on import guidelines, mention lazy loading

### PR DESCRIPTION
**Pull request by tupui**
Now that we have lazy imports (well since a few releases), we really can use things like ~`import scipy as sp`~ `import scipy`. Some of our dependent such as NetworkX even updated their whole codebase (not sure if finished, but I know they started) to take advantage of that.

Here I just propose to adjust our recommendations about import style and show that it's possible. I am not going down the road of recommending anything, simply showing this is possible and adjusting the wording around that.

### CODE REVIEW

File: doc/source/reference/index.rst [hunk 1]
```diff
@@ -41,25 +41,26 @@ to import modules/functions/objects from SciPy.
 Guidelines for importing functions from SciPy
 ---------------------------------------------
 
-The scipy namespace itself only contains functions imported from numpy.  These
+The SciPy namespace itself only contains functions imported from NumPy. These
```
Comment (rossbar) [inline doc/source/reference/index.rst:44]: I'm not sure this opening paragraph is strictly true anymore, at least from the user's perspective as defined by `__dir__`:

```python
>>> >>> import scipy
>>> dir(scipy)
['LowLevelCallable', '__numpy_version__', '__version__', 'cluster', 'datasets', 'fft', 'fftpack', 'integrate', 'interpolate', 'io', 'linalg', 'misc', 'ndimage', 'odr', 'optimize', 'show_config', 'signal', 'sparse', 'spatial', 'special', 'stats', 'test']
>>> scipy.__version__
'1.10.1'
```

Comment (tupui) [reply doc/source/reference/index.rst:44]: So it's a bit weird but we sadly still have some NumPy stuff 😓 Like

```python
import scipy as sp
sp.rand
# <function RandomState.rand>
```

We do raise a warning now if you try to use these. I am happy to remove the mention at all. But some folks here prefer that we describe exactly the current state of things vs what we want people to know.

Comment (rossbar) [reply doc/source/reference/index.rst:44]: Gotcha - this is just one area where the module `__dir__` and `__getattr__` muddy the waters about what "public" means.

Comment (rgommers) [reply doc/source/reference/index.rst:44]: On the actual diff: "the scipy namespace" refers to a code object (a namespace), so lower-case seems right. Maybe add backticks instead to typeset as code?

**Review by rgommers — CHANGES_REQUESTED** [review]
I'm sorry, I'm -1 on starting with `import scipy as sp`  or really mentioning that as an abbreviation at all. It just muddy the waters, and since there's very little in the main namespace and no precedent, starting to use `sp` is more confusing that helpful.

**Comment by tupui** [comment]
There is actually a precedent. We used to have that in our very own doc for quite a while (at least 8 years):

In 0.7 in 2008
https://github.com/scipy/scipy/commit/3f374927ec7b7bd4392a41040c28b450a2db434d

Removed in 2014:
https://github.com/scipy/scipy/pull/3933

And according to sites like this one https://import-as.github.io/import/scipy.html that's what people use the most when they do that.

At the time, I suppose the removal was made to avoid the import time issue.

**Comment by rkern** [comment]
The removal was made because it's introduction in 2008 was just a mistake that mostly went unnoticed because not many people ever used it (because just importing `scipy` and not its subpackages was useless at the time).

**Comment by tupui** [comment]
Can you elaborate on why this is a mistake given the current state of the library?

A lot of us think it is a good idea.

And also just from a public relation perspective/branding, that's a good idea. All packages followed NumPy with this aliasing idea.

File: doc/source/reference/index.rst (hunk 1 above)
Comment (rkern) [reply doc/source/reference/index.rst:44]: > But some folks here prefer that we describe exactly the current state of things vs what we want people to know.

Unfortunately, this is neither. Because of the lazy import mechanism, the namespace now includes the subpackages and does not "only contain the functions imported from `numpy`". To document the current state of things _and_ what we want people to know, we should elevate the lazy import mechanism up to the top here and relegate the mention that there are `numpy` functions down to the `.. note::`

**Comment by rkern** [comment]
Aliasing is always bad.

We do it out of necessity with `numpy` because we frequently need to write mathematical expressions with several things from the `numpy` namespace in one line; this is as much for the reader of the expression as the writer. The explosion of aliases for other packages that don't have the same needs is bad for programmer's brains trying to read code. We should always prioritize the reader of code over saving a few keystrokes for the writer of code.

`scipy` does not have the same needs as `numpy`. Because the `sp.` will always be followed by the name of the subpackage, there really is no appreciable savings here; it doesn't make any expressions more readable. If there actually is a need to abbreviate to fit many calls into an expression, then they'll need to `import as` the subpackage itself.

File: doc/source/reference/index.rst (hunk 1 above)
Comment (rkern) [reply doc/source/reference/index.rst:44]: You could just move this paragraph down to the end under a `.. note::` and remove "only".

**Comment by rkern** [comment]
IMO, a good test for whether or not a conventional alias is warranted is the one that primarily drove the `np` alias: In the absence of an alias, will people do `from package import *`? This was empirically the case for `Numeric` and early `numpy`. The `import Numeric as N` and `import numpy as np` aliases were direct responses to that common problematic practice. It was a compromise to avoid the worst alternative. In the absence of that worst alternative, I don't think we should make that compromise.

**Comment by tupui** [comment]
> will people do from package import *

Yes they do in our case since we have sub-packages. We do often things like `from scipy import stats` which is ok, but for something like `scipy.linalg`, this is a real concern as @rossbar demonstrated on the mailing list. Quoting him:

> For example, in networkx we have references to at least 4 different
`linalg` packages: `numpy.linalg`, `scipy.linalg`, `scipy.sparse.linalg`
and `networkx.linalg`. We now always refer to these explicitly from the
top-level package alias, i.e. `np.linalg`, `sp.linalg`, `sp.sparse.linalg`
etc.

To be clear, the text would actually not recommend to do anything. It is saying that thing are possible. If the aliasing is really the issue, fine, I will bite the bullet and propose `import scipy`. I will just say that I find this discussion ridiculous now.

**Comment by rkern** [comment]
I meant `*` literally, not as a dummy to refer to imports like `from scipy import linalg`. In the bad old days, people would write code that looked like this:

```python
from numpy import *

x = linspace(0, 1, 11)
y = cos(x)
```

*That* was the practice that we needed to forestall with a short alias for `numpy`.

**Comment by tupui** [comment]
Ok yes `form ... import *` is arguably a bad practice. Still having a namespace such as `linalg` is also very bad since it could mean 4 different things.

But again, this PR' intent is not saying you SHOULD/MUST use `import scipy as sp` instead of `from scipy import something`. It's just saying you can do both and removes the "negative" wording around importing the whole library.

As I said, if you and Ralf are really agains giving SciPy an alias, I will remove it.

**Comment by rkern** [comment]
No one is against talking about `import scipy` here. It should indeed be discussed here. It's only the `sp` alias that we are requesting to change.

**Comment by stefanv** [comment]
Reading Robert's last comment, it sounds like there is agreement on the benefits of importing the "root namespace":

1. One import statement at the top requires a lot less jumping around in code while coding to add further import statements.
2. Users never need to guess which subpackage (of numpy? of networkx? of scipy?) a function comes from, since it is spelled out: `scipy.linalg.x`.
3. Newcomers who rely on tab completion to discover functionality can start exploring SciPy with a single import.

W.r.t. the alias, whether we recommend for it or not, people seem to like having shorthands like `np` and `sp`. Especially if you embrace the "single import" paradigm, then having to type `sp.linalg.norm` is more convenient than `scipy.linalg.norm`, and I suspect people will keep doing it.

A long time ago, the pattern used to be `import numpy as N` until `np` was recommended. Similarly, there are now many versions of `scipy` shorthands, such as `import scipy.linalg as sla`; a recommendation from the project will at least help to standardize the convention.

**Comment by rkern** [comment]
I think this PR should go out with `import scipy` and no mention of aliases (either for or against or neutral). The addition of the lazy import mechanism makes the factual statements in the `main` version of this text out of date, and this PR is a welcome addition to fix that. Because Ralf's and my objections to the `sp` alias got confused for objections to mentioning the consequences of the lazy import mechanism, I propose that we move the alias discussion elsewhere.

**Comment by stefanv** [comment]
That makes sense, and the update without the alias would be a helpful improvement.

**Comment by tupui** [comment]
I made the change and remove the alias.

**Review by stefanv — APPROVED** [review]

File: doc/source/reference/index.rst [hunk 2]
```diff
@@ -88,6 +84,12 @@ section that the submodule in question is public. Of course you can still use::
 .. note:: SciPy is using a lazy loading mechanism which means that modules
           are only loaded in memory when you first try to access them.
 
+.. note::
+
+    The ``scipy`` namespace itself only contains functions imported from NumPy.
+    These functions still exist for backwards compatibility, but should be
+    imported from NumPy directly.
```
Comment (rkern) [inline doc/source/reference/index.rst:91]:
```suggestion
    The ``scipy`` namespace itself also contains functions imported from ``numpy``.
    These functions still exist for backwards compatibility, but should be
    imported from ``numpy`` directly.
```

**Comment by stefanv** [comment]
Thanks, Pamphile, for addressing all comments above.

File: doc/source/reference/index.rst [hunk 3]
```diff
@@ -41,25 +41,22 @@ to import modules/functions/objects from SciPy.
...
   result = optimize.curve_fit(...)
 
-This form of importing submodules is preferred for all submodules except
-``scipy.io`` (because ``io`` is also the name of a module in the Python
-stdlib)::
-
-  from scipy import interpolate
-  from scipy import integrate
-  import scipy.io as spio
+.. warning:: For ``scipy.io`` prefer the use of  ``import scipy``
```
Comment (rgommers) [inline doc/source/reference/index.rst:57]: This shouldn't be an admonition I think, certainly not a warning. Can you change to either `.. note::` or a plan sentence?

**Review by rgommers — COMMENTED** [review]
Thanks @tupui and all. One more comment on the changes from me, then this seems good to merge.

> it sounds like there is agreement on the benefits of importing the "root namespace":

Let me rephrase a little: it's fine to use `import scipy` as an equal alternative. It's not clearly better, it's just a tradeoff - sometimes it helps with clarity, at the cost of more characters. I'm fine with presenting it first as done here. 

What I would like to avoid seeing as follow-ups is changing the style of our docstring examples, they're now nicely uniform and it'd be a lot of churn to change that for little benefit.

**Review by rgommers — APPROVED** [review]
Thanks @tupui & all - the last open comment was addressed and this now seems fine with everyone, so in it goes.
